src/hotspot/share/gc/g1/g1FullCollector.cpp

```diff
@@ -350,35 +350,52 @@ bool G1FullCollector::phase2b_forward_oops() {
 
 void G1FullCollector::phase2c_prepare_serial_compaction() {
   GCTraceTime(Debug, gc, phases) debug("Phase 2: Prepare serial compaction", scope()->timer());
-  // At this point we know that after parallel compaction there will be no
-  // completely free regions. That means that the last region of
-  // all compaction queues still have data in them. We try to compact
-  // these regions in serial to avoid a premature OOM when the mutator wants
-  // to allocate the first eden region after gc.
+  // At this point, we know that after parallel compaction there will be regions that
+  // are partially compacted into. Thus, the last compaction region of all
+  // compaction queues still have space in them. We try to re-compact these regions
+  // in serial to avoid a premature OOM when the mutator wants to allocate the first
+  // eden region after gc.
+
+  // For maximum compaction, we need to re-prepare all objects above the lowest
+  // region among the current regions for all thread compaction points. It may
+  // happen that due to the uneven distribution of objects to parallel threads, holes
+  // have been created as threads compact to different target regions between the
+  // lowest and the highest region in the tails of the compaction points.
+  uint lowest_current = (uint)-1;
   for (uint i = 0; i < workers(); i++) {
     G1FullGCCompactionPoint* cp = compaction_point(i);
     if (cp->has_regions()) {
-      serial_compaction_point()->add(cp->remove_last());
+      lowest_current = MIN2(lowest_current, cp->current_region()->hrm_index());
     }
   }
 
-  // Update the forwarding information for the regions in the serial
-  // compaction point.
-  G1FullGCCompactionPoint* cp = serial_compaction_point();
-  for (GrowableArrayIterator<HeapRegion*> it = cp->regions()->begin(); it != cp->regions()->end(); ++it) {
-    HeapRegion* current = *it;
-    if (!cp->is_initialized()) {
-      // Initialize the compaction point. Nothing more is needed for the first heap region
-      // since it is already prepared for compaction.
-      cp->initialize(current);
-    } else {
-      assert(!current->is_humongous(), "Should be no humongous regions in compaction queue");
-      G1SerialRePrepareClosure re_prepare(cp, current);
-      set_compaction_top(current, current->bottom());
-      current->apply_to_marked_objects(mark_bitmap(), &re_prepare);
+  for (uint i = 0; i < workers(); i++) {
+    G1FullGCCompactionPoint* cp = compaction_point(i);
+    if (cp->has_regions()) {
+      cp->remove_at_or_above(lowest_current);
+    }
+  }
+
+  G1FullGCCompactionPoint* serial_cp = serial_compaction_point();
+  HeapWord* dense_prefix_top = nullptr;
+  for (uint i = lowest_current; i < _heap->max_reserved_regions(); i++) {
+    if (_region_attr_table.is_free(i) || _region_attr_table.is_compacting(i)) {
```

Comment:
```suggestion
    if (is_compaction_target(i)) {
```
This predicate is already available.

Reply:
Thanks, I missed that.